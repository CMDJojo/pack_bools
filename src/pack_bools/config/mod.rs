use proc_macro2::TokenStream;
use quote::quote;
use syn::Visibility as Vis;
use syn::VisRestricted;

pub use global::Config as GlobalConfig;
pub use global::GenType;
pub use local::Config as LocalConfig;

use crate::pack_bools::config::global::VisibilityTemplate;
use crate::pack_bools::config::local::{Accessor, VisibilityIdent};

mod global;
mod local;
pub mod parse;

#[derive(Debug)]
pub enum Visibility {
    Inherit,
    Public,
    Private,
    Restricted(VisRestricted),
}

impl Visibility {
    pub fn to_visibility(&self, inherited: &Vis) -> TokenStream {
        match self {
            Visibility::Inherit => quote! {
                #inherited
            },
            Visibility::Public => quote! {
                pub
            },
            Visibility::Private => TokenStream::new(),
            Visibility::Restricted(res) => quote! {
                #res
            },
        }
    }
}

pub struct CombinedConfig<'a> {
    global: &'a GlobalConfig,
    local: &'a LocalConfig,
}

impl<'a> CombinedConfig<'a> {
    pub fn new(global: &'a GlobalConfig, local: &'a LocalConfig) -> Self {
        Self { global, local }
    }

    pub fn setter(&self, ident: &str, vis: &Vis) -> Option<TokenStream> {
        match &self.local.setter {
            Accessor::None => None,
            Accessor::Custom(custom) => {
                let VisibilityIdent { visibility, ident } = custom;
                let vis = visibility.to_visibility(vis);
                Some(quote! {#vis fn #ident})
            }
            Accessor::Default => match &self.global.setter {
                None => None,
                Some(template) => {
                    let VisibilityTemplate {
                        visibility,
                        template,
                    } = template;
                    let vis = visibility.to_visibility(vis);
                    let name: TokenStream = template.format(ident).parse().unwrap();
                    Some(quote! {#vis fn #name})
                }
            },
        }
    }

    pub fn getter(&self, ident: &str, vis: &Vis) -> Option<TokenStream> {
        match &self.local.getter {
            Accessor::None => None,
            Accessor::Custom(custom) => {
                let VisibilityIdent { visibility, ident } = custom;
                let vis = visibility.to_visibility(vis);
                Some(quote! {#vis fn #ident})
            }
            Accessor::Default => match &self.global.getter {
                None => None,
                Some(template) => {
                    let VisibilityTemplate {
                        visibility,
                        template,
                    } = template;
                    let vis = visibility.to_visibility(vis);
                    let name: TokenStream = template.format(ident).parse().unwrap();
                    Some(quote! {#vis fn #name})
                }
            },
        }
    }
}
