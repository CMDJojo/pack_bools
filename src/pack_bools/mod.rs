use std::iter;

use proc_macro2::TokenStream;
use quote::{format_ident, quote, ToTokens};
use syn::{Error, Field, ItemStruct, Meta, parse2, Type};
use syn::spanned::Spanned;

use crate::pack_bools::config::{CombinedConfig, GenType, GlobalConfig, LocalConfig};

pub mod config;

macro_rules! try_syn {
    ($e:expr) => {
        match $e {
            Err(err) => return err.to_compile_error(),
            Ok(ok) => ok,
        }
    };
}

#[allow(clippy::too_many_lines)]
pub fn pack_bools(config: GlobalConfig, definition: ItemStruct) -> TokenStream {
    let ItemStruct {
        attrs,
        vis,
        struct_token,
        ident,
        generics,
        fields,
        semi_token: _,
    } = definition;
    let modified: Result<Vec<_>, _> = fields.into_iter().map(ModifiedField::from_field).collect();
    let modified = try_syn!(modified);

    let (to_convert, to_keep): (Vec<_>, Vec<_>) =
        modified.into_iter().partition(ModifiedField::is_packed);

    let to_keep: Option<Vec<_>> = to_keep
        .into_iter()
        .map(ModifiedField::into_excluded)
        .collect();

    let to_keep = to_keep.unwrap();
    let new_field: TokenStream = config.field_name.0.parse().unwrap();
    let inner_type = try_syn!(config
        .packed_type
        .to_packed_type(to_convert.len(), ident.span()));
    let (packed_type, newtype) = match &config.gen_type {
        GenType::Inline => (inner_type.to_token_stream(), None),
        GenType::NewType(name) => {
            let newtype_name = if let Some(name) = name {
                name.parse().unwrap()
            } else {
                format_ident!("{ident}PackedBools").to_token_stream()
            };
            let typedef = quote! {
                #[derive(Copy, Clone, Debug)]
                #[repr(transparent)]
                struct #newtype_name (#inner_type);
            };
            (newtype_name, Some(typedef))
        }
    };

    let packed_path = if config.gen_type.is_inline() {
        new_field.clone()
    } else {
        quote! { #new_field.0 }
    };

    let mut default_bits = 0u128;
    let mut functions = vec![];
    for (idx, (field, local)) in to_convert
        .into_iter()
        .map(|f| {
            let (a, b) = f.into_packaged().unwrap();
            (a, b.unwrap_or_default())
        })
        .enumerate()
    {
        if local.default {
            if config.gen_type.is_inline() {
                return Error::new(field.span(), "#[pack_bools(default = true)] only available on `newtype` generation: use #[pack_bools(newtype)] on the struct").to_compile_error();
            }
            default_bits |= 1 << idx;
        }

        let idx = idx.to_token_stream();
        let combined = CombinedConfig::new(&config, &local);
        if let Some(getter) = combined.getter(
            field.ident.as_ref().unwrap().to_string().as_str(),
            &field.vis,
        ) {
            let getter = quote! {
                #getter (&self) -> bool {
                    self. #packed_path & 1 << #idx != 0
                }
            };
            functions.push(getter);
        }

        if let Some(setter) = combined.setter(
            field.ident.as_ref().unwrap().to_string().as_str(),
            &field.vis,
        ) {
            let setter = quote! {
                #setter (&mut self, value: bool) {
                    let val = self. #packed_path;
                    self.#packed_path = if value {
                        val | 1 << #idx
                    } else {
                        val & !(1 << #idx)
                    };
                }
            };
            functions.push(setter);
        }
    }

    let default_impl = config.gen_type.is_newtype().then(|| {
        quote! {
            impl ::std::default::Default for #packed_type {
                fn default() -> Self {
                    Self( #default_bits as #inner_type )
                }
            }
        }
    });

    let (impl_generics, type_generics, where_clause) = generics.split_for_impl();

    let fields = to_keep
        .into_iter()
        .map(Field::into_token_stream)
        .chain(iter::once(quote! { #new_field: #packed_type }));

    let output = quote! {
        #(#attrs)* #vis #struct_token #ident #generics {
            #(#fields),*
        }

        #newtype

        #default_impl

        impl #impl_generics #ident #type_generics #where_clause {
            #(#functions)*
        }
    };

    output
}

pub enum ModifiedField {
    Excluded(Field),
    PackedField(Field, Option<LocalConfig>),
}

impl ModifiedField {
    pub fn is_packed(&self) -> bool {
        matches!(self, ModifiedField::PackedField(_, _))
    }

    pub fn into_excluded(self) -> Option<Field> {
        match self {
            ModifiedField::Excluded(e) => Some(e),
            ModifiedField::PackedField(_, _) => None,
        }
    }

    pub fn into_packaged(self) -> Option<(Field, Option<LocalConfig>)> {
        match self {
            ModifiedField::PackedField(f, c) => Some((f, c)),
            ModifiedField::Excluded(_) => None,
        }
    }
}

impl ModifiedField {
    pub fn from_field(field: Field) -> syn::Result<Self> {
        let mut config: Option<LocalConfig> = None;
        let Field {
            attrs,
            vis,
            mutability,
            ident,
            colon_token,
            ty,
        } = field;

        let mut new_attributes = Vec::with_capacity(attrs.len());
        for attr in attrs {
            if attr.path().is_ident("pack_bools") {
                // Remove this attribute
                if !is_bool_type(&ty) {
                    return Err(Error::new(
                        attr.span(),
                        "#[pack_bools] can only be used on bools",
                    ));
                } else if config.is_some() {
                    return Err(Error::new(
                        attr.span(),
                        "At most one #[pack_bools] attribute allowed per field",
                    ));
                }
                match attr.meta {
                    Meta::List(list) => {
                        let pc = parse2(list.tokens)?;
                        config = Some(pc);
                    }
                    _ => {
                        return Err(Error::new(
                            attr.span(),
                            "Use #[pack_bools(...)] for configuring fields",
                        ))
                    }
                }
            } else {
                // Keep other attributes
                new_attributes.push(attr);
            }
        }

        let is_bool = is_bool_type(&ty);
        let field = Field {
            attrs: new_attributes,
            vis,
            mutability,
            ident,
            colon_token,
            ty,
        };

        let res = if let Some(config) = config {
            if config.skip {
                Self::Excluded(field)
            } else {
                Self::PackedField(field, Some(config))
            }
        } else if is_bool {
            Self::PackedField(field, None)
        } else {
            Self::Excluded(field)
        };
        Ok(res)
    }
}

fn is_bool_type(ty: &Type) -> bool {
    if let Type::Path(p) = ty {
        p.path.is_ident("bool")
    } else {
        false
    }
}
