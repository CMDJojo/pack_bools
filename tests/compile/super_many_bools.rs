use pack_bools::pack_bools;

#[pack_bools(type = u8)]
struct MyBools {
    b1: bool,
    b2: bool,
    b3: bool,
    b4: bool,
    b5: bool,
    b6: bool,
    b7: bool,
    b8: bool,
    b9: bool,
    b10: bool,
    b11: bool,
    b12: bool,
    b13: bool,
    b14: bool,
    b15: bool,
    b16: bool,
    b17: bool,
    b18: bool,
    b19: bool,
    b20: bool,
    b21: bool,
    b22: bool,
    b23: bool,
    b24: bool,
    b25: bool,
    b26: bool,
    b27: bool,
    b28: bool,
    b29: bool,
    b30: bool,
    b31: bool,
    b32: bool,
    b33: bool,
    b34: bool,
    b35: bool,
    b36: bool,
    b37: bool,
    b38: bool,
    b39: bool,
    b40: bool,
    b41: bool,
    b42: bool,
    b43: bool,
    b44: bool,
    b45: bool,
    b46: bool,
    b47: bool,
    b48: bool,
    b49: bool,
    b50: bool,
    b51: bool,
    b52: bool,
    b53: bool,
    b54: bool,
    b55: bool,
    b56: bool,
    b57: bool,
    b58: bool,
    b59: bool,
    b60: bool,
    b61: bool,
    b62: bool,
    b63: bool,
    b64: bool,
    b65: bool,
    b66: bool,
    b67: bool,
    b68: bool,
    b69: bool,
    b70: bool,
    b71: bool,
    b72: bool,
    b73: bool,
    b74: bool,
    b75: bool,
    b76: bool,
    b77: bool,
    b78: bool,
    b79: bool,
    b80: bool,
    b81: bool,
    b82: bool,
    b83: bool,
    b84: bool,
    b85: bool,
    b86: bool,
    b87: bool,
    b88: bool,
    b89: bool,
    b90: bool,
    b91: bool,
    b92: bool,
    b93: bool,
    b94: bool,
    b95: bool,
    b96: bool,
    b97: bool,
    b98: bool,
    b99: bool,
    b100: bool,
    b101: bool,
    b102: bool,
    b103: bool,
    b104: bool,
    b105: bool,
    b106: bool,
    b107: bool,
    b108: bool,
    b109: bool,
    b110: bool,
    b111: bool,
    b112: bool,
    b113: bool,
    b114: bool,
    b115: bool,
    b116: bool,
    b117: bool,
    b118: bool,
    b119: bool,
    b120: bool,
    b121: bool,
    b122: bool,
    b123: bool,
    b124: bool,
    b125: bool,
    b126: bool,
    b127: bool,
    b128: bool,
    b129: bool,
}

fn main() {}
