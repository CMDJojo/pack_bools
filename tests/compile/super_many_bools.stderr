error: This struct contains more bools than would fit in your specified bit width
 --> tests/compile/super_many_bools.rs:4:8
  |
4 | struct MyBools {
  |        ^^^^^^^
